### ISSUE TITLE
Build inconsistencies with meson test

### ISSUE THREAD

**Issue by mikkelfj** [mesonbuild/meson#4902]
When running `meson test 'name of my test'`, the build does not always reflect reality using
meson 0.47.2, ninja 1.8.2, ccache 3.4.2, clang-1000.11.45., MacOS

The test is a basic C program linking with a basic C library under test.

ccache is enabled.

I have been been developing while testing and building using the test to drive the build.
Several times I have been debugging hard without understanding the problem, giving up, returning half a day later, and suddenly the problem has gone. On other occasions I have forced a clean with ninja to resolve a problem. This could be a timeout in ccache, or it could be that I just did something different that triggered a proper build.

There isn't really anything magic other than like custom target and such, except for ccache.

It's hard to provide any demonstrable example as it would require my entire setup, and things have been working in the past.

**Comment by jpakkane** [comment]
When this happens does `ninja test` work? Does it do something different?

Also as an aside, you might try the brand new Ninja 1.9.0. It has bug fixes related to timestamps which might help with this issue.

**Comment by mikkelfj** [comment]
Yes, everything works, I set a break point before a failed test and debug with lldb, only the data I see makes no sense. I.e. taking the false branch when a value is obviously true.

I think on one occasion I only changed the integer value of a define in a directly included header file.

Thanks for the Ninja pointer.

**Comment by mikkelfj** [comment]
A bad formulation in original text: there are no cudtom targets here, only ccache.

**Comment by mikkelfj** [comment]
Ninja 1.9.0 does not resolve the issue.
The following does not rebuild

    ninja -C build test

but this does

    ninja -C build
    ninja -C build test

**Comment by jpakkane** [comment]
If you run `meson test <yourtest>` do you see it invoking Ninja? It should.

**Comment by mikkelfj** [comment]
Absolutely. I use ninja -C test and meson test a bit arbitrarily depending on where I am. BTW a meson -C option would be good to have. From memory the problem has been with the meson call, but it could well also happen with ninja.

**Comment by mikkelfj** [comment]
I just got ninja 1.9.0 but I have not yet worked with it along with meson to conclude anything.

**Comment by mikkelfj** [comment]
Anyway - I see where you are getting at. I'll keep an eye out on which systems are triggered when/if I encounter more problems.

**Comment by mikkelfj** [comment]
It is unlikely to be just cached header files (or, it would be a more than a week old cache resurfacing) because some missing function prototypes have been in the header for quite a while but then behave as if they have disappeared from the header. The original error report was related to some defined flags having a compiled value different from the one in the header file. This is rather confusing.

**Comment by bonzini** [comment]
Unclear what was happening and there's no reproducer. After 6 years it probably can be closed...
